### PR TITLE
Make random seeding consistent with PRNG keys everywhere instead of seeds

### LINKED ISSUE #172: Unifying passing PRNG seeds
Sometimes, we pass the `seed: int`, other times we directly use the `PRNGKeyArray`.
We should be more consistent with the `jax` ecosystem and pass the `PRNGKeyArray` directly in most cases.

### PR DESCRIPTION
**Pull request by othmanesebbouh**
This fixes #172, replacing all the `seed` with `rng`, setting `jax.random.PRNGKey(0)` as default where applicable.

closes #172

### CODE REVIEW

File: tests/solvers/quadratic/fgw_test.py [hunk 1]
```diff
@@ -250,7 +250,7 @@ def test_fgw_lr_memory(self, rng: jnp.ndarray, jit: bool):
   def test_fgw_lr_generic_cost_matrix(
       self, rng: jnp.ndarray, cost_rank: Union[int, Tuple[int, int, int]]
   ):
-    n, m = 70, 100
+    n, m = 20, 30
```
Comment (pierreablin) [inline tests/solvers/quadratic/fgw_test.py:253]: you should maybe raise an issue here to say that it bugs with bigger dims.

File: tests/tools/k_means_test.py [hunk 2]
```diff
@@ -170,8 +170,8 @@ def test_k_means_plus_plus_better_than_random(self, rng: jnp.ndarray):
     key1, key2 = jax.random.split(rng, 2)
     geom, _, _ = make_blobs(n_samples=50, centers=k, random_state=10)
 
-    res_random = k_means.k_means(geom, k, init="random", key=key1)
-    res_kpp = k_means.k_means(geom, k, init="k-means++", key=key2)
+    res_random = k_means.k_means(geom, k, init="random", rng=key1)
```
Comment (pierreablin) [inline tests/tools/k_means_test.py:173]: maybe change key1 to rng1 etc..

File: src/ott/problems/nn/dataset.py [hunk 3]
```diff
@@ -95,7 +95,7 @@ def create_sample_generators(self) -> Iterator[jnp.array]:
     Returns:
       A generator of samples from the Gaussian mixture.
     """
-    key = self.init_key
+    key = self.init_rng
```
Comment (pierreablin) [inline src/ott/problems/nn/dataset.py:98]: change name here to rng to be more consistent?

File: src/ott/problems/nn/dataset.py [hunk 4]
```diff
@@ -95,7 +95,7 @@ def create_sample_generators(self) -> Iterator[jnp.array]:
...
     while True:
       k1, k2, key = jax.random.split(key, 3)
```
Comment (pierreablin) [inline src/ott/problems/nn/dataset.py:100]: same here

File: src/ott/tools/k_means.py [hunk 5]
```diff
@@ -118,23 +118,23 @@ def _random_init(
...
     k: int,
-    key: jnp.ndarray,
+    rng: jnp.ndarray,
     n_local_trials: Optional[int] = None,
 ) -> jnp.ndarray:
 
-  def init_fn(geom: pointcloud.PointCloud, key: jnp.ndarray) -> KPPState:
-    key, next_key = jax.random.split(key, 2)
+  def init_fn(geom: pointcloud.PointCloud, rng: jnp.ndarray) -> KPPState:
+    key, next_key = jax.random.split(rng, 2)
```
Comment (pierreablin) [inline src/ott/tools/k_means.py:126]: same here

**Comment by marcocuturi** [comment]
thanks Othmane! you need to run `pre-commit` to get rid of linting issues, this is in the [CONTRIBUTING.md](https://github.com/ott-jax/ott/blob/main/CONTRIBUTING.md)

File: tests/solvers/quadratic/fgw_test.py (hunk 1 above)
Comment (michalk8) [reply tests/solvers/quadratic/fgw_test.py:253]: bugs -> doesn't converge

**Review by michalk8 — CHANGES_REQUESTED** [review]
Can you please also change the type hints in all cases tot `jax.random.PRNGKeyArray` to be consistent?
Ideally the hint should be changed in tests as well whenever the `rng: jnp.ndarray` fixture is used (see, e.g., [here](https://github.com/ott-jax/ott/blob/main/tests/solvers/linear/sinkhorn_test.py#L31)), this is not critical however.

File: src/ott/initializers/linear/initializers_lr.py [hunk 6]
```diff
@@ -66,7 +66,7 @@ def __init__(self, rank: int, **kwargs: Any):
   def init_q(
       self,
       ot_prob: Problem_t,
-      key: jnp.ndarray,
+      rng: jnp.ndarray,
```
[the remark this answers is not in the document]
Comment (pierreablin) [reply src/ott/initializers/linear/initializers_lr.py:69]: Same everywhere, it should always be `rng: jax.random.PRNGKeyArray` instead of `rng: jnp.ndarray`

**Comment by othmanesebbouh** [comment]
Yes, modified them and I’ll push the changes this afternoon

File: src/ott/problems/quadratic/quadratic_problem.py [hunk 7]
```diff
@@ -383,19 +383,19 @@ def convert(
       return self
 
     (geom_xx, geom_yy, geom_xy, *children), aux_data = self.tree_flatten()
-    (s1, s2, s3) = jax.random.split(jax.random.PRNGKey(seed), 3)[:, 0]
+    (k1, k2, k3) = jax.random.split(rng, 3)
```
Comment (pierreablin) [inline src/ott/problems/quadratic/quadratic_problem.py:386]:
```suggestion
    rng1, rng2, rng3 = jax.random.split(rng, 3)
```
(No parenthesis to be consistent, and k -> rng

File: src/ott/solvers/quadratic/gromov_wasserstein.py [hunk 8]
```diff
@@ -102,7 +102,7 @@ class GWState(NamedTuple):
       linearization of GW.
     linear_pb: Local linearization of the quadratic GW problem.
     old_transport_mass: Intermediary value of the mass of the transport matrix.
-    keys: Random keys passed to low-rank initializers at every GW iteration
+    rngs: Random keys passed to low-rank initializers at every GW iteration
```
Comment (pierreablin) [inline src/ott/solvers/quadratic/gromov_wasserstein.py:105]: why are there multiple keys here ?

Comment (pierreablin) [reply src/ott/solvers/quadratic/gromov_wasserstein.py:105]: @michalk8 wdyt?

File: src/ott/tools/k_means.py [hunk 9]
```diff
@@ -118,23 +118,23 @@ def _random_init(
...
-    return KPPState(key=next_key, centroids=centroids, centroid_dists=dists)
+    return KPPState(rng=next_key, centroids=centroids, centroid_dists=dists)
 
   def body_fn(
       iteration: int, const: Tuple[pointcloud.PointCloud, jnp.ndarray],
       state: KPPState, compute_error: bool
   ) -> KPPState:
     del compute_error
-    key, next_key = jax.random.split(state.key, 2)
+    key, next_key = jax.random.split(state.rng, 2)
```
Comment (pierreablin) [inline src/ott/tools/k_means.py:137]:
```suggestion
    rng, next_rng = jax.random.split(state.rng, 2)
```

File: src/ott/tools/k_means.py [hunk 10]
```diff
@@ -151,14 +151,14 @@ def body_fn(
...
-        key=next_key, centroids=centroids, centroid_dists=centroid_dists
+        rng=next_key, centroids=centroids, centroid_dists=centroid_dists
     )
 
   if n_local_trials is None:
     n_local_trials = 2 + int(math.log(k))
   assert n_local_trials > 0, n_local_trials
 
-  state = init_fn(geom, key)
+  state = init_fn(geom, rng)
```
Comment (pierreablin) [inline src/ott/tools/k_means.py:161]: here `rng` was not changed during the last `split` contrary to the code before the change, this might lead to a bug.

File: src/ott/tools/k_means.py [hunk 11]
```diff
@@ -401,9 +401,7 @@ def k_means(
     weights = jnp.ones(geom.shape[0])
   assert weights.shape == (geom.shape[0],)
 
-  if key is None:
-    key = jax.random.PRNGKey(0)
-  keys = jax.random.split(key, n_init)
+  keys = jax.random.split(rng, n_init)
```
Comment (pierreablin) [inline src/ott/tools/k_means.py:404]:
```suggestion
  rngs = jax.random.split(rng, n_init)
```

File: tests/solvers/quadratic/fgw_test.py [hunk 12]
```diff
@@ -268,7 +268,7 @@ def test_fgw_lr_generic_cost_matrix(
     lr_prob = prob.to_low_rank()
     assert lr_prob.is_low_rank
 
-    solver = gw_solver.GromovWasserstein(rank=5, epsilon=1.0)
+    solver = gw_solver.GromovWasserstein(rank=5, epsilon=10.0)
```
Comment (pierreablin) [inline tests/solvers/quadratic/fgw_test.py:271]: do you still need to increase epsilon with the changed dims?

File: src/ott/solvers/quadratic/gromov_wasserstein.py (hunk 8 above)
Comment (michalk8) [reply src/ott/solvers/quadratic/gromov_wasserstein.py:105]: The cleanest way would be to refactor and have only `key: Optional[jax.random.PRNGKeyArray]` that is being split in [iterations](https://github.com/ott-jax/ott/blob/main/src/ott/solvers/quadratic/gromov_wasserstein.py#L373) and passed onto the next state.

File: src/ott/tools/k_means.py (hunk 10 above)
Comment (michalk8) [reply src/ott/tools/k_means.py:161]: This should be fine, as apart from the `init_fn`, the `rng` is not used anywhere.

File: src/ott/solvers/quadratic/gromov_wasserstein.py (hunk 8 above)
Comment (pierreablin) [reply src/ott/solvers/quadratic/gromov_wasserstein.py:105]: You mean `rng: Optional[jax.random.PRNGKeyArray]` right? name should be rng?

File: tests/solvers/quadratic/fgw_test.py (hunk 12 above)
Comment (othmanesebbouh) [reply tests/solvers/quadratic/fgw_test.py:272]: I checked again, and for this specific seed yes, we need to increase epsilon, but for some others it's not necessary...

Comment (marcocuturi) [reply tests/solvers/quadratic/fgw_test.py:272]: sounds like a fairly steep change :) 

wouldn't it be more reasonable to change the tolerance? What kind of error is `epsilon=1.0` triggering?

Comment (othmanesebbouh) [reply tests/solvers/quadratic/fgw_test.py:272]: the algorithm doesn't converge (`assert out.converged` returns an `AssertionError`). I've just opened an issue to report this as a bug (#302).

**Comment by marcocuturi** [comment]
@othmanesebbouh do you think you could run the pre-commit to get rid of linting issues? Thanks!

**Comment by michalk8** [comment]
@othmanesebbouh any progress on this? Re the test that doesn't converge [here](https://github.com/ott-jax/ott/pull/290/files/74e0bbdbc1289a78c421b5a75a55b0417dc43bd1#r1108694791), I'd say either you can keep it as it is (high epsilon), or revert + use another seed (if it works). In any case, will look into it in #302

**Comment by othmanesebbouh** [comment]
@michalk8 just ran the pre-commit, I hope it works now. I actually thought I had already done that, but wasn't using it correctly, sorry.

For the test that doesn't converge, I feel that having epsilon very large is a less cryptic way to signaling that there's a problem than having a seed that miraculously works.

**Comment by michalk8** [comment]
> For the test that doesn't converge, I feel that having epsilon very large is a less cryptic way to signaling that there's a problem than having a seed that miraculously works.

Works for me, thanks! Could you please merge merge `main` and resolve the conflicts?
This can be done by, e.g.:
```bash
git remote add upstream ssh://git@github.com/ott-jax/ott
git fetch upstream
git merge upstream/main
```

**Review by michalk8 — APPROVED** [review]
Thanks a lot @othmanesebbouh, LGTM!
